### PR TITLE
Refactor JSON serialization and escape handling

### LINKED ISSUE #1682: Bug in nanoFramework.Json: First character of string value is lost when it starts with \u00XX
### Library/API/IoT binding

nanoFramework.Json

### Visual Studio version

VS2022

### .NET nanoFramework extension version

2022.14.1.2

### Target name(s)

ESP32-S3-WROOM

### Firmware version

N/A

### Device capabilities

_No response_

### Description

When deserializing a JSON string using `JsonConvert.DeserializeObject` in **.NET nanoFramework**, **the first character of a string value is completely dropped** if it is represented as a Unicode escape sequence `\u00XX` (e.g., `\u00225`, `\u00311`, `\u00412`, etc.) and located **at the beginning of the string value**.

---

### Example

**Input JSON (valid, with escaping):**
```json
{\u0022Product\u0022:\u00225 test\u0022,\u0022Count\u0022:10}
```

**Equivalent without escaping:**
```json
{"Product":"5 test","Count":10}
```

**Target class:**
```csharp
public class Item
{
    public string Product { get; set; }
    public int Count { get; set; }
}
```

**Deserialization code:**
```csharp
var obj = JsonConvert.DeserializeObject(json, typeof(Item)) as Item;
Debug.WriteLine(obj.Product);
```

---

### Expected Result  
```
5 test
```

---

### Actual Result  
```
 test
```

**The character `5` (from `\u00225`) is completely lost.**

---

### Conditions  
- The string value **starts with `\u00XX`**.
- `\u00XX` represents an **ASCII character** (digits `0`–`9`, letters `A`–`F`, etc.).
- The escaping is **valid and compliant** with the JSON standard.

### Root Cause  
The `nanoFramework.Json` parser:
- Processes `\uXXXX` sequences **before completing string literal parsing**.
- Treats `\u00XX` at the start of a value as a **separate token**.
- **Discards** the decoded character if it is the first one in the string.

This **violates the JSON standard** (RFC 8259), which requires `\uXXXX` inside strings to be properly decoded.

---

### Impact  
- **Data loss** when receiving JSON from external APIs, CMS, or cloud services.
- Inability to parse **valid but escaped** JSON.
- **Critical** for IoT devices where JSON is the primary data exchange format.

### Recommendation  
- **Fix the parser** in `nanoFramework.Json` to correctly handle `\uXXXX` **inside string values**.
- Add **unit tests** for Unicode escapes at the start of strings.
- Until fixed — **always apply preprocessing** when receiving escaped JSON.

---

**Status:** **Critical Bug**  
**Priority:** **High**  
**Repository:** `nanoframework/nanoFramework.Json`

### How to reproduce

_No response_

### Expected behaviour

_No response_

### Screenshots

_No response_

### Sample project or code

_No response_

### Aditional information

_No response_

### PR DESCRIPTION
**Pull request by networkfusion**
## Description

After parsing the original issue body, and then adding the latest comment,
Copilot said:

Thanks — that's a helpful additional repro. The symptom you show (Description == " liters" instead of "5 liters") is exactly the same class of bug: the initial character decoded from a leading \uXXXX escape is being dropped.

I updated GetNextTokenInternal previously to decode \uXXXX escapes and append the decoded character into the current string buffer. That change should address the case where a string value starts with a \uXXXX escape (for example "\u00315 liters") and prevent the first character from being dropped.

I can't run the nanoFramework test harness here, but to validate locally you can add and run the small unit test below (or run the same deserialization code you posted). If you run it and still see the "5" missing, paste the exact JSON bytes you feed into Deserialize and I'll iterate further.

Suggested unit test (xUnit-like pseudocode — adapt to your test runner):
C#

``` csharp
// unit test for leading unicode escape in string value
public void Deserialize_StringStartingWithUnicodeEscape_ReturnsDecodedChar()
{
    // escaped JSON: "Product":"\u00315 liters" -> Product = "5 liters"
    var json = "{\"Product\":\"\\u00315 liters\",\"Count\":10}";

    var item = JsonConvert.DeserializeObject(json, typeof(Item)) as Item;

    Assert.NotNull(item);
    Assert.Equal("5 liters", item.Product);
}
```

If you prefer to test with the RpcResponse example you posted, construct a JSON where the nested "Description" starts with \u0031 and verify:
C#

``` csharp
var message = "{\"Result\":\"{\\\"ProductId\\\":1,\\\"Name\\\":\\\"A 5-liter bottle\\\",\\\"Description\\\":\\\"\\u00315 liters\\\",\\\"Price\\\":100,\\\"LiterCount\\\":5}\"}";
var responseRpc = JsonConvert.DeserializeObject(message, typeof(RpcResponse)) as RpcResponse;
Debug.WriteLine($"Field result:{responseRpc.Result}");
// Then parse responseRpc.Result as JSON to inspect Description
```

What to do next if the test still fails

    If the "5" is still dropped, please paste the exact JSON you pass to JsonConvert.DeserializeObject (including backslashes as in the raw string literal) or the byte[] content. I will use that exact input to adjust the parser handling.
    If the test passes, I will prepare a small PR adding the unit test(s) and the fix so it can be merged.

Recap of the fix I proposed

    Decode exactly 4 hex digits after a '\u' and append the decoded char into the active string buffer, advancing jsonPos correctly.
    Preserve handling for other escapes (\n, \t, ", \, etc.).
    Add unit tests for strings beginning with \uXXXX and for \uXXXX appearing in the middle of strings.

## Motivation and Context

- Fixes nanoFramework/Home#1682

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: nanoFramework.Json.Test/JsonSerializerOptionsTests.cs [hunk 1]
```diff
@@ -119,6 +119,32 @@ public void Can_serialize_and_deserialize_arrays_of_class_objects()
...
+
+
+            JsonTestCompany deserializedResult = (JsonTestCompany)JsonConvert.DeserializeObject(jsonString, typeof(JsonTestCompany));
+
+            Assert.NotNull(deserializedResult);
+            Assert.Equal("5 Guys", deserializedResult.CompanyName);
+            Assert.Equal(10, deserializedResult.CompanyID);
+
+            OutputHelper.WriteLine("Finished test...");
+        }
```
[the remark this answers is not in the document]
Comment (networkfusion) [reply nanoFramework.Json.Test/JsonSerializerOptionsTests.cs:146]: Is this not the point that we are trying to check?
